src/npyffi/mod.rs

```diff
@@ -10,13 +10,21 @@
 )]
 
 use std::mem::forget;
+use std::os::raw::c_uint;
 use std::os::raw::c_void;
 
 use pyo3::{
     types::{PyAnyMethods, PyCapsule, PyCapsuleMethods, PyModule},
     PyResult, Python,
 };
 
+#[cfg(not(any(feature = "numpy-1", feature = "numpy-2")))]
+compile_error!("at least one of feature \"numpy-1\" and feature \"numpy-2\" must be enabled");
+
+pub const NPY_2_0_API_VERSION: c_uint = 0x00000012;
+
+pub static ABI_API_VERSIONS: std::sync::OnceLock<(c_uint, c_uint)> = std::sync::OnceLock::new();
```

Comment:
I think `GILOnceCell` is applicable here as we can just require that the accessor functions like `PyDataType_FLAGS` take a `py: Python` token. (We define them here so there is not need to conform exactly to the C signature just as there is no guarantee that they will stay in sync with the definitions in NumPy's C headers.)

Reply:
Besides accessor functions, the version is also needed in the API functions that have a different offset in Numpy 1 and 2 (for instance `PyArray_CopyInto`) or only exist in Numpy 1 or 2 (using them with the wrong runtime version would otherwise result in memory corruption or a segfault).

I'm happy to switch to `GILOnceCell` but just wanted to check that changing higher-level function signatures was ok. For instance, https://github.com/PyO3/rust-numpy/blob/2170e16bc5eedaa969d9da732d470f788d725020/src/dtype.rs#L259 will need `py: Python` so that it can call `PyDataType_FLAGS`).

Reply:
If you have GIL ref `&'py PyArray<..>` or a bound ref `Bound<'py, PyArray<..>>`, this implies access to a GIL token via e.g. [`Bound::py`](https://docs.rs/pyo3/latest/pyo3/prelude/struct.Bound.html#method.py). So I don't think this is an issue.

In any case, this will be a breaking release so we can change the API where necessary.